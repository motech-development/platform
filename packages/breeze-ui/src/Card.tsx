import React, { FC } from 'react';
import styled from 'styled-components';

const BaseCard = styled.div``;

const Card: FC = ({ children }) => {
  return <BaseCard>{children}</BaseCard>;
};

export default Card;
