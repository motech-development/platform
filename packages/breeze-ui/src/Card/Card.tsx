import React, { FC, ReactNode } from 'react';
import styled, { ThemeProvider } from 'styled-components';

type CardPadding = 'sm' | 'md' | 'lg';

interface ICardTheme {
  [name: string]: {
    padding: string;
  };
}

const cardTheme: ICardTheme = {
  lg: {
    padding: '20px',
  },
  md: {
    padding: '10px',
  },
  sm: {
    padding: '0',
  },
};

interface IBaseCard {
  padding: CardPadding;
}

const BaseCard = styled.div<IBaseCard>`
  ${({ theme, padding }) => `
    background: #fff;
    padding: ${theme[padding].padding}
  `}
`;

export interface ICardProps {
  children: ReactNode;
  padding?: CardPadding;
}

const Card: FC<ICardProps> = ({ children, padding = 'sm' }) => {
  return (
    <ThemeProvider theme={cardTheme}>
      <BaseCard padding={padding}>{children}</BaseCard>
    </ThemeProvider>
  );
};

export default Card;
