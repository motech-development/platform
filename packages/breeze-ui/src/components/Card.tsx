import { FC, ReactNode } from 'react';
import { classNames, spacingClass, themeClass } from '../utils/className';
import TSpacing from '../utils/spacing';
import TTheme from '../utils/theme';

export interface ICardProps {
  children: ReactNode;
  className?: string;
  flex?: boolean;
  padding?: TSpacing;
  theme?: TTheme;
}

const Card: FC<ICardProps> = ({
  children,
  className = '',
  flex = false,
  padding = 'md',
  theme = 'secondary',
}) => (
  <>
    {/* @tailwind: bg-blue-100 border-blue-300 text-blue-800 */}
    {/* @tailwind: bg-gray-100 border-gray-300 text-gray-800 */}
    {/* @tailwind: bg-green-100 border-green-300 text-green-800 */}
    {/* @tailwind: bg-red-100 border-red-300 text-red-800 */}
    {/* @tailwind: bg-yellow-100 border-yellow-300 text-yellow-800 */}
    {/* @tailwind: p-0 p-2 p-4 p-6 */}
    <div
      className={classNames(
        'border-b-2 shadow-lg',
        className,
        flex ? 'flex flex-col items-center' : '',
        spacingClass(padding, 'p-{spacing}'),
        themeClass(theme, 'bg-{theme}-100 border-{theme}-300 text-{theme}-800'),
      )}
    >
      {children}
    </div>
  </>
);

export default Card;
