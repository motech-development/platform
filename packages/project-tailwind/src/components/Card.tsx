import { ComponentPropsWithoutRef, forwardRef } from 'react';
import {
  Sizing,
  Themes,
  TSizing,
  TTheme,
  useTailwind,
} from '../utilities/tailwind';

/** Card component props */
export interface ICardProps extends ComponentPropsWithoutRef<'div'> {
  /** Apply flexbox styles */
  flex?: boolean;

  /** Component padding */
  padding?: TSizing;

  /** Componenet theme */
  theme?: TTheme;
}

/**
 * Flexible and extensible content container
 *
 * @param props - Component props
 *
 * @returns Card componenet
 */
export const Card = forwardRef<HTMLDivElement, ICardProps>(
  (
    {
      className,
      flex = false,
      padding = Sizing.MD,
      theme = Themes.SECONDARY,
      ...rest
    },
    ref,
  ) => {
    const { createStyles } = useTailwind(theme, padding);

    const cardStyles = createStyles({
      classNames: [
        'border-b-2 shadow-lg',
        {
          'flex flex-col items-center': flex,
        },
        className,
      ],
      sizing: {
        lg: ['p-6'],
        md: ['p-4'],
        none: ['p-0'],
        sm: ['p-2'],
      },
      theme: {
        danger: ['bg-red-100 border-red-300 text-red-800'],
        primary: ['bg-blue-100 border-blue-300 text-blue-800'],
        secondary: ['bg-gray-100 border-gray-300 text-gray-800'],
        success: ['bg-green-100 border-green-300 text-green-800'],
        warning: ['bg-yellow-100 border-yellow-300 text-yellow-800'],
      },
    });

    return <div className={cardStyles} ref={ref} {...rest} />;
  },
);
