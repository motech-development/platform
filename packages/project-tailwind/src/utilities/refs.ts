import { LegacyRef, MutableRefObject, useMemo, useRef } from 'react';

/**
 * Check if ref is a mutable ref object
 *
 * @param ref - Reference object or function
 *
 * @returns If ref is a mutable ref object
 */
function isMutableRefObject<T>(
  ref: MutableRefObject<T> | LegacyRef<T>,
): ref is MutableRefObject<T | null> {
  return ref != null;
}

/** useMergeRefs arguments */
type TUseMergeRefsInput<T> = Array<MutableRefObject<T> | LegacyRef<T>>;

/**
 * Combines multiple refs into a single memoized callback instance
 *
 * @param refs - Reference object or function
 *
 * @returns Combined ref object
 */
export default function useMergeRefs<T>(...refs: TUseMergeRefsInput<T>) {
  const targetRef = useRef<T>(null);

  return useMemo(() => {
    if (refs.every((ref) => ref == null)) {
      return null;
    }

    refs.forEach((ref) => {
      if (typeof ref === 'function') {
        ref(targetRef.current);
      } else if (isMutableRefObject(ref)) {
        // eslint-disable-next-line no-param-reassign
        ref.current = targetRef.current;
      }
    });

    return targetRef;
  }, [refs]);
}
