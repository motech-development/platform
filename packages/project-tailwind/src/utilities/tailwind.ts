import clsx, { ClassValue } from 'clsx';

/** Alignment options */
export enum Alignment {
  /** Centre */
  CENTRE = 'centre',
  /** Left */
  LEFT = 'left',
  /** Right */
  RIGHT = 'right',
}

/** Alignment type */
export type TAlignment = `${Alignment}`;

/** Sizing options */
export enum Sizing {
  /** Large */
  LG = 'lg',
  /** Medium */
  MD = 'md',
  /** None */
  NONE = 'none',
  /** Small */
  SM = 'sm',
}

/** Sizing type */
export type TSizing = `${Sizing}`;

/** Theming options */
export enum Themes {
  /** Danger */
  DANGER = 'danger',
  /** Primary */
  PRIMARY = 'primary',
  /** Secondary */
  SECONDARY = 'secondary',
  /** Success */
  SUCCESS = 'success',
  /** Warning */
  WARNING = 'warning',
}

/** Theme type */
export type TTheme = `${Themes}`;

/** Class names options */
type TCreateStylesClassNames = {
  /** Universal classes to apply */
  classNames?: ClassValue[];
  /** Sizing specific classes */
  sizing?: {
    [name in Sizing]?: ClassValue[];
  };
  /** Theme specific classes */
  theme?: {
    [name in TTheme]?: ClassValue[];
  };
};

/**
 * Tailwind hook
 *
 * @param theme - The theme to apply
 * @param sizing - Sizing to apply
 *
 * @returns Tailwind helper utilities
 */
export function useTailwind(theme: TTheme, sizing: TSizing) {
  /**
   * Creates css styles to apply
   *
   * @param classNames - Classes to be applied
   *
   * @returns Class names
   */
  function createStyles(classNames: TCreateStylesClassNames) {
    let finalStyles: ClassValue[] = [];

    if (classNames.classNames) {
      finalStyles = [...classNames.classNames];
    }

    if (sizing && classNames.sizing) {
      const styles = classNames.sizing[sizing];

      if (!styles) {
        throw new Error('No classes defined for this sizing.');
      }

      finalStyles = [...finalStyles, ...styles];
    }

    if (theme && classNames.theme) {
      const styles = classNames.theme[theme];

      if (!styles) {
        throw new Error('No classes defined for this theme.');
      }

      finalStyles = [...finalStyles, ...styles];
    }

    return clsx(finalStyles);
  }

  /**
   * Creates text alignment classes
   *
   * @param alignment - Text alignment
   *
   * @returns Class names
   */
  function createTextAlignmentStyles(alignment: TAlignment) {
    switch (alignment) {
      case Alignment.CENTRE:
        return 'text-center';
      case Alignment.RIGHT:
        return 'text-right';
      case Alignment.LEFT:
      default:
        return 'text-left';
    }
  }

  return {
    createStyles,
    createTextAlignmentStyles,
  };
}
