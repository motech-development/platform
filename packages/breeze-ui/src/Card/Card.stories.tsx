import { select, text, withKnobs } from '@storybook/addon-knobs';
import BaseStyles from '../BaseStyles/BaseStyles';
import Card from './Card';

export default {
  component: Card,
  decorators: [withKnobs],
};

const padding = {
  Large: 'lg',
  Medium: 'md',
  Small: 'sm',
};

export const BasicCard = {
  name: 'Basic card',
  render: () => (
    <>
      <BaseStyles />

      <Card padding={select('Padding', padding, 'md') as 'sm' | 'md' | 'lg'}>
        {text('Text', 'Hello world')}
      </Card>
    </>
  ),
};
