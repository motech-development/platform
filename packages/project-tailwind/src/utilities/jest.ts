/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderResult } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UserEvent } from '@testing-library/user-event/dist/types/setup/setup';
import { ReactElement } from 'react';
import { Sizing, Themes } from './tailwind';

export const themes = [
  {
    theme: Themes.DANGER,
  },
  {
    theme: Themes.PRIMARY,
  },
  {
    theme: Themes.SECONDARY,
  },
  {
    theme: Themes.SUCCESS,
  },
  {
    theme: Themes.WARNING,
  },
];

export const sizing = [
  {
    size: Sizing.LG,
  },
  {
    size: Sizing.MD,
  },
  {
    size: Sizing.NONE,
  },
  {
    size: Sizing.SM,
  },
];

type TSetup = RenderResult & {
  user: UserEvent;
};

/**
 * Combines Testing Library render function with user events setup
 *
 * @param ui - React element to test
 *
 * @returns Rendered component and test user
 */
export function setup(ui: ReactElement): TSetup {
  return {
    user: userEvent.setup({
      delay: null,
    }),
    ...render(ui),
  };
}
