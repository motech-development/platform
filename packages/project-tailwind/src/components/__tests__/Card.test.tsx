import { setup, sizing, themes } from '../../utilities/jest';
import { Card } from '../Card';

describe('Card', () => {
  describe.each(themes)('when theme is "$theme"', ({ theme }) => {
    it('should render the correct output', () => {
      const { asFragment } = setup(
        <Card theme={theme}>
          <p>Hello, world</p>
        </Card>,
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should render the correct output when flex is set', () => {
      const { asFragment } = setup(
        <Card flex theme={theme}>
          <p>Hello, world</p>
        </Card>,
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it.each(sizing)(
      'should render the correct output when padding is "$size"',
      ({ size }) => {
        const { asFragment } = setup(
          <Card theme={theme} padding={size}>
            <p>Hello, world</p>
          </Card>,
        );

        expect(asFragment()).toMatchSnapshot();
      },
    );
  });
});
