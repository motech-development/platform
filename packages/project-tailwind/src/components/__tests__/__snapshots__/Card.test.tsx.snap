// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card when theme is "danger" should render the correct output 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-4 bg-red-100 border-red-300 text-red-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "danger" should render the correct output when flex is set 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg flex flex-col items-center p-4 bg-red-100 border-red-300 text-red-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "danger" should render the correct output when padding is "lg" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-6 bg-red-100 border-red-300 text-red-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "danger" should render the correct output when padding is "md" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-4 bg-red-100 border-red-300 text-red-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "danger" should render the correct output when padding is "none" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-0 bg-red-100 border-red-300 text-red-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "danger" should render the correct output when padding is "sm" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-2 bg-red-100 border-red-300 text-red-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "primary" should render the correct output 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-4 bg-blue-100 border-blue-300 text-blue-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "primary" should render the correct output when flex is set 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg flex flex-col items-center p-4 bg-blue-100 border-blue-300 text-blue-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "primary" should render the correct output when padding is "lg" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-6 bg-blue-100 border-blue-300 text-blue-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "primary" should render the correct output when padding is "md" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-4 bg-blue-100 border-blue-300 text-blue-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "primary" should render the correct output when padding is "none" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-0 bg-blue-100 border-blue-300 text-blue-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "primary" should render the correct output when padding is "sm" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-2 bg-blue-100 border-blue-300 text-blue-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "secondary" should render the correct output 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-4 bg-gray-100 border-gray-300 text-gray-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "secondary" should render the correct output when flex is set 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg flex flex-col items-center p-4 bg-gray-100 border-gray-300 text-gray-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "secondary" should render the correct output when padding is "lg" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-6 bg-gray-100 border-gray-300 text-gray-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "secondary" should render the correct output when padding is "md" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-4 bg-gray-100 border-gray-300 text-gray-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "secondary" should render the correct output when padding is "none" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-0 bg-gray-100 border-gray-300 text-gray-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "secondary" should render the correct output when padding is "sm" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-2 bg-gray-100 border-gray-300 text-gray-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "success" should render the correct output 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-4 bg-green-100 border-green-300 text-green-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "success" should render the correct output when flex is set 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg flex flex-col items-center p-4 bg-green-100 border-green-300 text-green-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "success" should render the correct output when padding is "lg" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-6 bg-green-100 border-green-300 text-green-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "success" should render the correct output when padding is "md" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-4 bg-green-100 border-green-300 text-green-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "success" should render the correct output when padding is "none" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-0 bg-green-100 border-green-300 text-green-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "success" should render the correct output when padding is "sm" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-2 bg-green-100 border-green-300 text-green-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "warning" should render the correct output 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-4 bg-yellow-100 border-yellow-300 text-yellow-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "warning" should render the correct output when flex is set 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg flex flex-col items-center p-4 bg-yellow-100 border-yellow-300 text-yellow-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "warning" should render the correct output when padding is "lg" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-6 bg-yellow-100 border-yellow-300 text-yellow-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "warning" should render the correct output when padding is "md" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-4 bg-yellow-100 border-yellow-300 text-yellow-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "warning" should render the correct output when padding is "none" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-0 bg-yellow-100 border-yellow-300 text-yellow-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;

exports[`Card when theme is "warning" should render the correct output when padding is "sm" 1`] = `
<DocumentFragment>
  <div
    class="border-b-2 shadow-lg p-2 bg-yellow-100 border-yellow-300 text-yellow-800"
  >
    <p>
      Hello, world
    </p>
  </div>
</DocumentFragment>
`;
